modules/net/include/hephaestus/net/recv.h

```diff
@@ -0,0 +1,138 @@
+//=================================================================================================
+// Copyright (C) 2023-2025 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <cstddef>
+#include <span>
+#include <system_error>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#include <liburing.h>  // NOLINT(misc-include-cleaner)
+#include <liburing/io_uring.h>
+#include <stdexec/__detail/__execution_fwd.hpp>
+#include <stdexec/__detail/__senders_core.hpp>
+#include <stdexec/execution.hpp>
+
+#include "hephaestus/concurrency/basic_sender.h"
+#include "hephaestus/net/detail/operation_state.h"
+#include "hephaestus/net/socket.h"
+
+namespace heph::net {
+template <bool SendAll>
+struct RecvT {
+  template <stdexec::sender Sender>
+  auto operator()(Sender&& sender, Socket const& socket, std::span<std::byte> buffer) const {
+    auto domain = stdexec::__get_early_domain(sender);
+    return stdexec::transform_sender(domain,
+                                     concurrency::makeSenderExpression<RecvT>(std::tuple{ &socket, buffer },
+                                                                              std::forward<Sender>(sender)));
+  }
+  template <stdexec::sender Sender>
+  auto operator()(Sender&& sender, Socket const* socket, std::span<std::byte> buffer) const {
+    auto domain = stdexec::__get_early_domain(sender);
+    return stdexec::transform_sender(domain, concurrency::makeSenderExpression<RecvT>(
+                                                 std::tuple{ socket, buffer }, std::forward<Sender>(sender)));
+  }
+
+  auto operator()(Socket const& socket, std::span<std::byte> buffer) const
+      -> stdexec::__binder_back<RecvT, Socket const*, std::span<std::byte>> {
+    return { { &socket, buffer }, {}, {} };
+  }
+};
+
+// NOLINTNEXTLINE (readability-identifier-naming)
+static inline RecvT<false> recv{};
```

Comment:
consider using normal functions

Reply:
unfortunately not possible due to the partial application of the function and the desired `operator|` overload.